### PR TITLE
Week4 jason wang

### PR DESCRIPTION
**Pull request by hawk198723**

### CODE REVIEW

**Comment by hawk198723** [comment]
those 4 pushes are all the same. because I had some issues when I pushed to the first branch "week-4-jason-wang", I had some errors below:
error: src refspec week4-jaosn-wang does not match any
error: failed to push some refs to 'https://github.com/learn-academy-2021-echo/week-4-assessment-hawk198723.git'

Then I have to make another branch, and then push it again. and Find out there are 4 pushes already.

File: code-challenges.rb [hunk 1]
```diff
@@ -10,6 +10,18 @@
...
+    if num % 2 == 1
+         "#{num} is odd."
+    elsif num % 2 ==0
+         "#{num} is even."
+    else
+         "Please input a integer."
+    end
+end
+    p is_odd_or_even num2
+
```
Comment (sjproctor) [inline code-challenges.rb:23]: Nice work. Make sure you are testing your code with all the given inputs.

File: code-challenges.rb [hunk 2]
```diff
@@ -20,8 +32,14 @@
...
 # Expected output: 'bby Rd'
+ 
 
-
+def removes_all_vowels str
+    newstr = str.delete "aeiouAEIOU"
+    newstr
+end
+    p removes_all_vowels album3
+
```
Comment (sjproctor) [inline code-challenges.rb:42]: Looks great. I think you can leave out the `newstr` variable and just let the implicit return handle the logic.

File: code-challenges.rb [hunk 3]
```diff
@@ -31,3 +49,13 @@
...
+def palindrome?(string)
+    if string.downcase == string.reverse.downcase
+       p "#{string} is a palindrome."
+    else
+       p "#{string} is not a palindrome."
+    end
+end
+
+palindrome? is_palindrome3
```
Comment (sjproctor) [inline code-challenges.rb:61]: I would be cautious about defining your method with a question mark. Another best practice is to make sure the `p` is on the method invocation.

File: code-challenges.test.js [hunk 4]
```diff
@@ -20,7 +20,38 @@ var colors1 = ["purple", "blue", "green", "yellow", "pink"]
...
+    let colors2 = ["chartreuse", "indigo", "periwinkle", "ochre", "aquamarine", "saffron"]
+    // a test/it method, nested within the describe block, that in plain words, describes that the function does.
+    it("takes in an array, removes the first item from the array and shuffles the remaining content.", () => {
+
+        //an expect method, nested within the test block, calling on the shuffleTheRest() function, followed by the .toEqual() matcher that checks the expected output of the function return.
+        expect(shuffleTheRest(colors1)).toEqual(expect.arrayContaining(["yellow", "blue", "pink", "green"]))
+        expect(shuffleTheRest(colors2)).toEqual(expect.arrayContaining(["saffron", "aquamarine", "periwinkle", "indigo", "ochre"]))
+    })
+})
+// -- Got the red --
```
Comment (sjproctor) [inline code-challenges.test.js:35]: Nice job with the test.

File: code-challenges.test.js [hunk 5]
```diff
@@ -20,7 +20,38 @@ var colors1 = ["purple", "blue", "green", "yellow", "pink"]
...
+// console.log(shuffleTheRest(colors1));
+
+// -- Got the green -- 
+
+// -- Refactor --  
+
+function shuffleTheRest(array) {
+    array.shift();
+    return shuffledArray = array.sort((a, b) => 0.5 - Math.random()); 
+}
```
Comment (sjproctor) [inline code-challenges.test.js:54]: I like the refactor. In this case, I don't think you need to pass any arguments to the sort method.

File: code-challenges.test.js [hunk 6]
```diff
@@ -29,18 +60,40 @@ var colors2 = ["chartreuse", "indigo", "periwinkle", "ochre", "aquamarine", "saf
...
+//     //return the array
+//     return minAndMaxArray
+// }
+// -- Got the green -- 
 
+// -- Refactor --  
+function minAndMax(array) {
 
+    return minAndMaxArray = [Math.min(...array), Math.max(...array)]
+}
```
Comment (sjproctor) [inline code-challenges.test.js:96]: Beautiful! Great work.

File: code-challenges.test.js [hunk 7]
```diff
@@ -49,6 +102,26 @@ var testArray1 = [3, 7, 10, 5, 4, 3, 3]
...
+//     // declare the array and use spread operator to conbine the arrays and use [...new Set []] to take out the duplicate values
+//     let newArray =  [...new Set([...array1, ...array2])]
+//     //return newArray 
+//     return newArray
+// }
+// -- Refactor --  
+function noDuplicate(array1, array2) {
+    
+    return newArray =  [...new Set([...array1, ...array2])]
+}
```
Comment (sjproctor) [inline code-challenges.test.js:127]: Nice work. I think the only thing would be removing the newArray. `return [...new Set([...array1, ...array2])]`

File: interview-questions.md [hunk 8]
```diff
@@ -7,52 +7,63 @@ Then, research the question to expand on your answer. Even if you feel you have
...
+touch events ( TouchEvent ): touchstart, touchmove, touchend, touchcancel.
+keyboard events ( KeyboardEvent ): keydown, keypress, keyup.
+form events: focus, blur, change, submit.
 
 
 
 3. What is object-oriented programming? How is it different than functional programming?
 
-  Your answer:
+  Your answer: it's a very abstrct programming language. You can use it to build an encapsulate abstrct object. Whenever we need it, use can just call that object's name(instance).  Functional programming , such C language. we cann't do it like OOP language, the data has to run from top to buttom. Not that flexible like OOP language.
```
Comment (sjproctor) [inline interview-questions.md:35]: Nice answer.

**Review by sjproctor — APPROVED** [review]
Really great job, Jason. Your JS code looks great. Keep an eye out for those times when you can streamline your code by removing excess variable declarations.

**Pull request merged**